groups/voteForDelegate.js

```diff
@@ -0,0 +1,116 @@
+const axios = require('axios');
+const get = require('./get');
+const logger = require('../helpers/logger');
+const keys = require('../helpers/keys');
+const constants = require('../helpers/constants');
+const transactionFormer = require('../helpers/transactionFormer');
+const validator = require('../helpers/validator');
+
+const DEFAULT_VOTE_FOR_DELEGATE_RETRIES = 4; // How much re-tries for send tokens requests by default. Total 4+1 tries
+
+const publicKeysCache = { };
+
+module.exports = (nodeManager) => {
+  /**
+		* Creates votes for delegate transaction, signs it, and broadcasts to ADAMANT network
+    * See https://github.com/Adamant-im/adamant/wiki/Transaction-Types#type-3-vote-for-delegate-transaction
+    * @param {string} passPhrase Senders's passPhrase. Sender's address will be derived from it.
+    * @param {string[]} votes PublicKeys, ADM addresses and delegate names for upvote and downvote.
+    * It would be more efficient to pass publicKey, otherwise the api will make additional queries
+    * @param {number} maxRetries How much times to retry request
+    * @returns {Promise} Request results
+  	*/
+	return async (passPhrase, votes, maxRetries = DEFAULT_VOTE_FOR_DELEGATE_RETRIES, retryNo = 0) => {
+
+    let transaction;
+
+    try {
+      if (!validator.validatePassPhrase(passPhrase))
+			  return validator.badParameter('passPhrase');
+
+      const keyPair = keys.createKeypairFromPassPhrase(passPhrase);
+
+      const uniqueVotes = [];
+
+      for (let i = votes.length - 1; i >= 0; i--) {
+        const vote = votes[i];
+        const voteName = vote.slice(1);
+        const voteDirection = vote.charAt(0);
```

Comment:
You can use votes.reverse().forEach((votes, i) => {...}) instead

Reply:
I think current solution is preferred, as `forEach` is async function, and we need sync cycle

Reply:
But we can do like 
`await Promise.all(votes.reverse().map((votes, i) => {...})) `
anyway, pure for-cycle may be cleaner for async logic inside, you can decide what is better

Reply:
So we can mark as solved.